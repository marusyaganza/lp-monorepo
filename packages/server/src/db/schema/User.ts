import { Schema, model } from 'mongoose';
import { User as UserCoreType } from 'generated/graphql';

export interface UserType extends UserCoreType {
  password: string;
  primaryLanguage?: string;
}

const userSchema = new Schema<UserType>({
  firstName: { type: String, required: true },
  lastName: { type: String, required: true },
  email: { type: String, required: true, unique: true },
  createdAt: { type: String, required: true },
  role: { type: String, enum: ['ADMIN', 'GUEST', 'MEMBER'], required: true },
  password: { type: String, required: true },
  primaryLanguage: String
});

export const User = model<UserType>('User', userSchema);
