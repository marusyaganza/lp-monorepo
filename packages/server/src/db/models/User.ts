import { User, UserType } from '../schema/User';
import { SignUpInput, Role } from '../../generated/graphql';
import { formatData, formatFilter } from '../helpers';

export interface UserModelType {
  findOne: (filter: Partial<UserType>) => Promise<UserType | null>;
  findMany: (filter: Partial<UserType>) => Promise<UserType[] | null>;
  createOne: (fields: SignUpInput & { role: Role }) => Promise<UserType | null>;
  updateOne: (
    fields: Partial<UserType>
  ) => Promise<{ ok: boolean; value: UserType | null }>;
  deleteOne: (filter: { id: string }) => Promise<{ ok: boolean }>;
}

export const UserModel: UserModelType = {
  async findOne(filter = {}) {
    const user = await User.findOne(formatFilter(filter));
    return formatData(user);
  },

  //TODO check this method when admin auth is ready
  async findMany(filter = {}) {
    const users = await User.find(formatFilter(filter));
    return users;
  },

  async createOne(fields) {
    const createdAt = Date.now();
    const user = await User.create({ ...fields, createdAt });
    return formatData(user);
  },

  //TODO check this method when GQL for this is ready
  async updateOne(fields) {
    const update = { ...fields };
    delete update.id;
    const { ok, value } = await User.findByIdAndUpdate(fields.id, update, {
      includeResultMetadata: true,
      new: true
    });
    return { ok: Boolean(ok), value: formatData(value) };
  },

  //TODO check this method when admin auth is ready
  async deleteOne(filter) {
    const result = { ok: false };
    const { deletedCount, acknowledged } = await User.deleteOne(
      formatFilter(filter)
    );
    if (acknowledged && deletedCount == 1) {
      result.ok = true;
    }
    return result;
  }
};
