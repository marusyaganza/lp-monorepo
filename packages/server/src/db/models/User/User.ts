import { User, UserType } from '../../schema/User';
import { SignUpInput } from '../../../generated/graphql';
import {
  AuthenticationError,
  OperationResolutionError
} from '../../../utils/apolloCustomErrors';
import { DEFAULT_ROLE } from '../../../constants/defaultValues';
import { hashPassword } from '../../../auth';
import { ERROR_MESSAGES } from '../../../constants/errorMessages';

export interface UserModelType {
  findOne: (filter: Partial<UserType>) => Promise<UserType | null | undefined>;
  createOne: (fields: SignUpInput) => Promise<UserType | null>;
  createTempUser: () => Promise<UserType | null>;
}

export const UserModel: UserModelType = {
  async findOne(filter = {}) {
    const { id } = filter;
    let user;
    if (id) {
      user = await User.findById(id);
    } else {
      user = await User.findOne(filter);
    }
    return user?.toObject();
  },

  async createOne(fields) {
    const { email, password } = fields;
    const existing = await User.findOne({ email });
    if (existing) {
      throw new AuthenticationError(`user with email ${email} already exists`);
    }
    const role = DEFAULT_ROLE;
    const hashedPassword = await hashPassword(password);
    if (!hashedPassword) {
      throw new OperationResolutionError(ERROR_MESSAGES.SIGN_UP_FAILED);
    }
    const user = await User.create({
      ...fields,
      role,
      password: hashedPassword
    });
    return user.toObject();
  },
  async createTempUser() {
    const role = DEFAULT_ROLE;
    const timestamp = Date.now();
    const firstName = 'Sam';
    const lastName = 'Smith';
    const email = `${firstName.toLowerCase()}.${lastName.toLowerCase()}.${timestamp}@example.com`;
    const hashedPassword = await hashPassword(timestamp.toString());
    if (!hashedPassword) {
      throw new OperationResolutionError(ERROR_MESSAGES.SIGN_UP_FAILED);
    }
    const user = await User.create({
      firstName,
      lastName,
      email,
      role,
      password: hashedPassword
    });
    return user.toObject();
  }
};
