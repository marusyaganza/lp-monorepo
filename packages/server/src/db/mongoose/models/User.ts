import { Types, Document } from 'mongoose';
import { User } from '../User';
import { User as UserType } from '../../../generated/graphql';
import { ModelType } from '../../';

type DocumentType =
  | (Document<unknown, {}, UserType> &
      Omit<
        UserType & {
          _id: Types.ObjectId;
        },
        never
      >)
  | null;

function formatData(data: DocumentType) {
  if (!data) {
    return null;
  }
  return data.toObject({ getters: true });
}

export const UserModel: ModelType<UserType> = {
  async findOne(filter = {}) {
    const user = await User.findById(filter.id);
    return formatData(user);
  },

  //TODO check this method when admin auth is ready
  async findMany(filter = {}) {
    const users = await User.find(filter);
    return users;
  },

  async createOne(fields) {
    const createdAt = Date.now();
    const user = new User({ ...fields, createdAt });
    await user.save();
    // const user = await User.create({...fields, createdAt});
    // console.log('user', user.toObject({ getters: true }));
    return formatData(user);
  },

  //TODO check this method when GQL for this is ready
  async updateOne(fields) {
    const update = { ...fields };
    delete update.id;
    const { ok, value } = await User.findByIdAndUpdate(fields.id, update, {
      rawResult: true
    });
    return { ok: Boolean(ok), value };
  },

  //TODO check this method when admin auth is ready
  async deleteOne(filter) {
    const result = await User.deleteOne(filter);
    console.log('result', result);
    return { ok: true };
  }
};
