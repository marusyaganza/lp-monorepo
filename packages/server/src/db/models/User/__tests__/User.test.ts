import { UserModel } from '../User';
import {
  connectToDb,
  disconnectFromDb,
  dropDb,
  baseSnapshotConfig
} from '../../../../tests/helpers';

const snapshotConfig = {
  ...baseSnapshotConfig,
  createdAt: expect.any(String)
};

const createUserInput = {
  email: 'test@test.com',
  password: 'password1',
  firstName: 'Test',
  lastName: 'User1',
  primaryLanguage: 'English',
  role: 'MEMBER'
};

describe('User', () => {
  beforeEach(async () => {
    await connectToDb();
    await dropDb();
  });

  afterEach(async () => {
    await dropDb();
    await disconnectFromDb();
  });

  test('createOne', async () => {
    const result = await UserModel.createOne(createUserInput);
    expect(result).toMatchSnapshot(snapshotConfig);
  });

  test('findOne', async () => {
    const user = await UserModel.createOne(createUserInput);
    const result = await UserModel.findOne({
      email: createUserInput.email
    });
    const result2 = await UserModel.findOne({ id: user?.id });
    expect(result2).not.toBeNull();
    expect(result).toMatchSnapshot(snapshotConfig);
  });
});
