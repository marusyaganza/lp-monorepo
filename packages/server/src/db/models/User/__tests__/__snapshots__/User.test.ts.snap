// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User createOne 1`] = `
{
  "_id": Any<Object>,
  "createdAt": Any<String>,
  "email": "test@test.com",
  "firstName": "Test",
  "id": Any<String>,
  "lastName": "User1",
  "password": "mock_hash_password1_salt_5",
  "primaryLanguage": "English",
  "role": "MEMBER",
}
`;

exports[`User findOne 1`] = `
{
  "_id": Any<Object>,
  "createdAt": Any<String>,
  "email": "test@test.com",
  "firstName": "Test",
  "id": Any<String>,
  "lastName": "User1",
  "password": "mock_hash_password1_salt_5",
  "primaryLanguage": "English",
  "role": "MEMBER",
}
`;
